docker/creds/credentials.go

```diff
@@ -0,0 +1,451 @@
+		opts.promptForCredentials = cbk
+	}
+}
+
+// WithVerifyCredentials sets custom callback for credentials validation.
+func WithVerifyCredentials(cbk VerifyCredentialsCallback) Opt {
+	return func(opts *credentialsProvider) {
+		opts.verifyCredentials = cbk
+	}
+}
+
+// WithPromptForCredentialStore sets custom callback that is supposed to
+// interactively ask user which credentials store/helper is used to store credentials obtained
+// from user.
+func WithPromptForCredentialStore(cbk ChooseCredentialHelperCallback) Opt {
+	return func(opts *credentialsProvider) {
+		opts.promptForCredentialStore = cbk
+	}
+}
+
+func WithTransport(transport http.RoundTripper) Opt {
+	return func(opts *credentialsProvider) {
+		opts.transport = transport
+	}
+}
+
+// WithAdditionalCredentialLoaders adds custom callbacks for credential retrieval.
+// The callbacks are supposed to be non-interactive as opposed to WithPromptForCredentials.
+//
+// This might be useful when credentials are shared with some other service.
+//
+// Example: OpenShift builtin registry shares credentials with the cluster (k8s) credentials.
+func WithAdditionalCredentialLoaders(loaders ...CredentialsCallback) Opt {
+	return func(opts *credentialsProvider) {
+		opts.credentialLoaders = append(opts.credentialLoaders, loaders...)
+	}
+}
+
+// NewCredentialsProvider returns new CredentialsProvider that tires to get credentials from docker/func config files.
+//
+// In case getting credentials from the config files fails
+// the caller provided callback (see WithPromptForCredentials) will be invoked to obtain credentials.
+// The callback may be called multiple times in case the returned credentials
+// are not correct (see WithVerifyCredentials).
+//
+// When the callback succeeds the credentials will be saved by using helper defined in the func config.
+// If the helper is not defined in the config file
+// it may be picked by provided callback (see WithPromptForCredentialStore).
+// The picked value will be saved in the func config.
+//
+// To verify that credentials are correct custom callback can be used (see WithVerifyCredentials).
+func NewCredentialsProvider(opts ...Opt) docker.CredentialsProvider {
+	var c credentialsProvider
+
+	for _, o := range opts {
+		o(&c)
+	}
+
+	if c.transport == nil {
+		c.transport = http.DefaultTransport
+	}
+
+	if c.verifyCredentials == nil {
+		c.verifyCredentials = func(ctx context.Context, registry string, credentials docker.Credentials) error {
+			return CheckAuth(ctx, registry, credentials, c.transport)
+		}
+	}
+
+	if c.promptForCredentialStore == nil {
+		c.promptForCredentialStore = func(available []string) (string, error) {
+			return "", nil
+		}
+	}
+
+	c.authFilePath = filepath.Join(fn.ConfigPath(), "auth.json")
+	sys := &containersTypes.SystemContext{
+		AuthFilePath: c.authFilePath,
+	}
+
+	home, err := os.UserHomeDir()
+	if err != nil {
+		panic(err)
+		//return result, fmt.Errorf("failed to determine home directory: %w", err)
```

Comment:
?

Reply:
I must have forgotten this some times ago this is not new file.